fl4health/clients/fed_pca_client.py

```diff
@@ -0,0 +1,144 @@
+        # and as such, we will make that assumption.
+        self.num_train_samples = len(self.train_loader.dataset)  # type: ignore
+        self.num_val_samples = len(self.val_loader.dataset)  # type: ignore
+
+        self.initialized = True
+
+    def get_data_tensor(self, data_loader: DataLoader) -> Tensor:
+        raise NotImplementedError
+
+    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
+        """Perform PCA using the locally held dataset."""
+        if not self.initialized:
+            self.setup_client(config)
+        center_data = self.narrow_config_type(config, "center_data", bool)
+        train_data_tensor = self.get_data_tensor(self.train_loader)
+        train_data_tensor = train_data_tensor.to(self.device)
+
+        principal_components, singular_values = self.model(train_data_tensor, center_data)
+        self.model.set_principal_components(principal_components, singular_values)
+
+        metrics: Dict[str, Scalar] = {}
+        cumulative_explained_variance = self.model.compute_cumulative_explained_variance()
+        explained_variance_ratios = self.model.compute_explained_variance_ratios()
+        metrics["cumulative_explained_variance"] = cumulative_explained_variance
+        metrics["top_explained_variance_ratio"] = explained_variance_ratios[0].item()
+        return (self.get_parameters(config), self.num_train_samples, metrics)
+
+    def evaluate(self, parameters: NDArrays, config: Dict[str, Scalar]) -> Tuple[float, int, Dict[str, Scalar]]:
+        """
+        Evaluate merged principal components on the local validation set.
+
+        Args:
+            parameters (NDArrays): Server-merged principal components.
+            config (Dict[str, Scalar]): Config file.
+        """
+        if not self.initialized:
+            self.setup_client(config)
+            train_data_tensor = self.get_data_tensor(self.train_loader)
+            self.model.set_data_mean(self.model.maybe_reshape(train_data_tensor))
+        self.set_parameters(parameters, config)
+        num_components_eval = (
+            self.narrow_config_type(config, "num_components_eval", int)
+            if "num_components_eval" in config.keys()
+            else None
+        )
+        val_data_tensor = self.get_data_tensor(self.val_loader)
+        val_data_tensor_prepared = self.model.centre_data(self.model.maybe_reshape(val_data_tensor)).to(self.device)
+        reconstruction_loss = self.model.compute_reconstruction_error(val_data_tensor_prepared, num_components_eval)
+        projection_variance = self.model.compute_projection_variance(val_data_tensor_prepared, num_components_eval)
+        metrics: Dict[str, Scalar] = {}
+        metrics["projection_variance"] = projection_variance
+        return (reconstruction_loss, self.num_val_samples, metrics)
+
+    def save_model(self) -> None:
+        final_model_save_path = f"{self.model_save_path}/client_{self.generate_hash()}_pca"
```

Comment:
We might want a file type at the end of this file name like `.pkl` or `cpt` or something?

Reply:
Yes, good point